src/common/AdpfWrapper.cpp

```diff
@@ -120,5 +120,22 @@ void AdpfWrapper::onEndCallback(double durationScaler) {
         int64_t actualDurationNanos = endCallbackNanos - mBeginCallbackNanos;
         int64_t scaledDurationNanos = static_cast<int64_t>(actualDurationNanos * durationScaler);
         reportActualDuration(scaledDurationNanos);
+        mPreviousDuration = scaledDurationNanos;
     }
 }
+
+void AdpfWrapper::reportWorkload(int32_t appWorkload) {
+    if (isOpen()) {
+        // Compare with previous workload. If we think we will need more
+        // time to render the callback then warn ADPF as early as possible.
+        if (appWorkload > mPreviousWorkload
+                && mPreviousWorkload > 0
+                && mPreviousDuration > 0) {
+            int64_t predictedDuration = appWorkload * mPreviousDuration / mPreviousWorkload;
+            reportActualDuration(predictedDuration);
```

Comment:
Could there be a race condition between the reportActualDuration in this function and the one in onEndCallback()?

Reply:
There is no race. ADPF will just have more information to work with. At the end of the callback, ADPF will get the actual duration, which may be affected by the CPU boost due to the last call.